subtests/docker_cli/login/login.py

```diff
@@ -0,0 +1,333 @@
+        self._check_credentials()
+
+
+class login_fail(login_base):
+
+    """
+    1. Run 'docker login' with incorrect password
+    2. Verify error exit status, and reason in stderr
+    """
+
+    def run_once(self):
+        super(login_fail, self).run_once()
+
+        wrong_passwd = '!' + self.sub_stuff['passwd']
+        self.sub_stuff["cmdresult"] = self.docker_login(wrong_passwd)
+
+    def postprocess(self):
+        super(login_fail, self).postprocess()
+        cmdresult = self.sub_stuff['cmdresult']
+        OutputGood(cmdresult, ignore_error=True)
+        mustfail(cmdresult, 1)
+        self.failif_not_in("no successful auth challenge", cmdresult.stderr,
+                           "stderr from failed docker login")
+        self.failif_not_in("401 Unauthorized", cmdresult.stderr,
+                           "stderr from failed docker login")
+
+
+class logout_ok(login_ok):
+
+    """
+    1. Run the login_ok test
+    2. Run docker logout
+    3. Make sure stored credentials are removed
+    4. Make sure stdout contains a logout message
+    """
+
+    def run_once(self):
+        super(logout_ok, self).run_once()
+        logout = DockerCmd(self, 'logout', [self.sub_stuff['server']])
+        self.sub_stuff['cmdresult_logout'] = logout.execute()
+
+    def postprocess(self):
+        super(logout_ok, self).postprocess()
+        self.failif_not_in('Remove login credentials for',
+                           self.sub_stuff['cmdresult_logout'].stdout,
+                           'stdout from docker logout')
+
+
+class push_ok(login_base):
+
+    """
+    1. Run 'docker login' with correct password
+    2. Run 'docker tag', then 'docker push' to the new local registry
+    3. Verify zero (success) exit status, and 'Pushed' in stdout
+    """
+
+    def run_once(self):
+        super(push_ok, self).run_once()
+        self.docker_login()
+        self.sub_stuff['pushresult'] = self.docker_push('push_ok', 'tag_ok')
+
+    def postprocess(self):
+        super(push_ok, self).postprocess()
+        cmdresult = self.sub_stuff['pushresult']
+        print cmdresult.stdout
+        OutputGood(cmdresult)
+        mustpass(cmdresult)
+        self.failif_not_in(": Pushed", cmdresult.stdout,
+                           "stdout from docker push")
+        self.failif_not_in("tag_ok: digest:", cmdresult.stdout,
+                           "stdout from docker push")
+
+
+class push_fail(login_base):
+
+    """
+    1. Run 'docker login' with incorrect password
+    2. Run 'docker tag', then 'docker push' to the new local registry
+    3. Verify error exit status
+    """
+
+    def run_once(self):
+        super(push_fail, self).run_once()
+        self.docker_login('!' + self.sub_stuff['passwd'])
+        self.sub_stuff['pushresult'] = self.docker_push('push_fail', 'tag_x')
+
+    def postprocess(self):
+        super(push_fail, self).postprocess()
+        cmdresult = self.sub_stuff['pushresult']
+        print cmdresult.stderr
```

Comment:
I wonder...because I make this same mistake as well as leaving being ``pdb.set_trace()`` sometimes.  Both would be good things to add to our checkers.  I'll make a note to take a stab at that.